### PR TITLE
nonoise and cleaning house

### PR DESCRIPTION
**Pull request by esheldon**
adapt test to new descwl-shear-sims sims

also beginning to remove old coadd code

### CODE REVIEW

**Comment by esheldon** [comment]
the `test_coadd_correct.py` needs to be adapted, but we no longer have SEObs or keep around galsim objects.

File: descwl_coadd/coadd.py [hunk 1]
```diff
@@ -1,1236 +1,639 @@
...
-        psf_data = self._make_warps(
-            exps=self.psf_exps,
-            dims=self.psf_dims,
-            wcs=self.coadd_psf_wcs,
-            dopsf=False,
-        )
+    coadd_psf_bbox = geom.Box2I(
+        geom.Point2I(0, 0),
+        geom.Point2I(psf_dims[0]-1, psf_dims[1]-1),
+    )
```
Comment (beckermr) [inline descwl_coadd/coadd.py:137]: Are bounding boxes supposed to start at zero or should this be centered on a stamp at the central pixel of the coadd?

File: descwl_coadd/coadd.py [hunk 2]
```diff
@@ -1,1236 +1,639 @@
...
 
-        psf_obj = self.coadd_exp.getPsf()
-        psf_image = psf_obj.computeKernelImage(stack_pos).array
+def get_exp_and_noise(exp_or_ref, rng, remove_poisson):
+    """
+    get the exposure (possibly from a deferred handle) and create
+    a corresponding noise exposure
 
-        psf_cen = (np.array(psf_image.shape)-1.0)/2.0
+    TODO move interpolating BRIGHT into metadetect or mdet-lsst-sim
```
Comment (beckermr) [inline descwl_coadd/coadd.py:261]: We cannot move this to the sim code. It needs to be in mdet or in the stack task that runs mdet on the coadds. This last option is the equivalent of where it is in DES.

File: descwl_coadd/coadd.py [hunk 3]
```diff
@@ -1,1236 +1,639 @@
...
+    coadd_wcs: DM wcs
+        The wcs for the coadd
+    var: float
+        The variance to set in the psf variance map
+
+    Returns
+    -------
+    psf ExposureF
+    """
+    coadd_sky_orig = coadd_wcs.getSkyOrigin()
```
Comment (beckermr) [inline descwl_coadd/coadd.py:540]: I guess this won't handle patches yet? I think each patch shares the overall WCS of the tract and is simple a subset of it?

File: descwl_coadd/coadd.py (hunk 2 above)
Comment (esheldon) [reply descwl_coadd/coadd.py:261]: mdet-lsst-sim isn't the sim code it is a code that runs metadetection on the sims

https://github.com/esheldon/mdet-lsst-sim

This is what I've used to run all the lsst tests

File: descwl_coadd/coadd.py (hunk 3 above)
Comment (esheldon) [reply descwl_coadd/coadd.py:540]: could be a bug

File: descwl_coadd/coadd.py (hunk 1 above)
Comment (esheldon) [reply descwl_coadd/coadd.py:137]: need to double check

File: descwl_coadd/tests/test_coadd.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+            coadd_bbox=sim_data['coadd_bbox'],
+            psf_dims=sim_data['psf_dims'],
+            rng=rng,
+            remove_poisson=False,  # no object poisson noise in sims
+        )
+
+        coadd_dims = (coadd_dim, )*2
+        psf_dims = (psf_dim, )*2
+        assert coadd.image.shape == coadd_dims
+        assert coadd.psf.image.shape == psf_dims
```
Comment (beckermr) [inline descwl_coadd/tests/test_coadd.py:99]: We should add some other tests here, like not all nan or something.

File: descwl_coadd/tests/test_coadd.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+
+def test_coadd_psgauss_smoke():
```
Comment (beckermr) [inline descwl_coadd/tests/test_coadd.py:102]: What does this test do that the previous one cannot?

File: descwl_coadd/tests/test_coadd.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+        assert coadd.image.shape == coadd_dims
+        assert coadd.noise.shape == coadd_dims
+        assert coadd.psf.image.shape == psf_dims
+
+
+@pytest.mark.skipif(
+    "CATSIM_DIR" not in os.environ,
+    reason='simulation input data is not present'
+)
+def test_coadd_pspsf_smoke():
```
Comment (beckermr) [inline descwl_coadd/tests/test_coadd.py:132]: This test like the one before it seems redundant.

**Review by beckermr — CHANGES_REQUESTED** [review]
I really like the new functional approach. It makes the code much more readable and easy to parse!

File: descwl_coadd/coadd.py (hunk 2 above)
Comment (beckermr) [reply descwl_coadd/coadd.py:261]: Right yes. In any case, I suspect we'll need to move the interp to the stack task that runs mdet.

Comment (esheldon) [reply descwl_coadd/coadd.py:261]: We need to have a single code path for this so we can test it once.  I don't want it to be both in the stack task and in a separate code we run on sims.

I think this leaves only metadetect as the place for it

File: descwl_coadd/coadd.py (hunk 1 above)
Comment (esheldon) [reply descwl_coadd/coadd.py:137]: I think zero is correct

File: descwl_coadd/tests/test_coadd.py (hunk 5 above)
Comment (esheldon) [reply descwl_coadd/tests/test_coadd.py:102]: this was a copy paste I forgot to fill in, removing

File: descwl_coadd/tests/test_coadd.py (hunk 6 above)
Comment (esheldon) [reply descwl_coadd/tests/test_coadd.py:132]: this one is power spectrum psf

File: descwl_coadd/tests/test_coadd.py (hunk 4 above)
Comment (esheldon) [reply descwl_coadd/tests/test_coadd.py:99]: done

File: descwl_coadd/tests/test_coadd.py (hunk 6 above)
Comment (beckermr) [reply descwl_coadd/tests/test_coadd.py:132]: But what does that have to do with the coadding? Simply putting that through is no different than the fixed PSFs.

File: descwl_coadd/coadd.py (hunk 3 above)
Comment (esheldon) [reply descwl_coadd/coadd.py]: I think this was a bug, pushed what I think is a fix

File: descwl_coadd/tests/test_coadd.py (hunk 6 above)
Comment (esheldon) [reply descwl_coadd/tests/test_coadd.py:108]: The underlying psf object from the sim works differently, wanted to make sure we could use it properly.

I guess its a test of the sim and that we can use the sim

If you object I can remove it

Comment (esheldon) [reply descwl_coadd/tests/test_coadd.py:108]: this is the only place where CATSIM_DIR is needed, that tips the scales, will remove it

**Comment by esheldon** [comment]
verified the current PR (with the corresponding PRs in descwl-shear-sims, metadetect) with grid tests

**Review by beckermr — CHANGES_REQUESTED** [review]
OK. I just realized what the code is doing. Right now when we get the PSF image for each SE exp, we are using a jacobian approximation to the full WCS at the PSF center. This is not what we should be doing. Instead, we should directly use the WCS of the SE exposure and attach that WCS to stamp cutout of the PSF around the location where the PSF center ends up. This means the stamp lower-left pixel will not be at index (0,0) nor will it match the patch index. Only this way of doing things will match the full WCS treatment we do for the images and noise images. 

We have not seen biases due to this mismatch now, but our sims are relatively simple in terms of their WCS features.

**Comment by esheldon** [comment]
I probably just don't understand all the operations I'm using.  The probably do not do what I think they are doing.

We should discuss offline with Eli and anyone else who might understand these things

**Comment by esheldon** [comment]
My attempt to restate Matt's proposal @erykoff @beckermr 

- create a psf image at location x, y and assign it the same wcs as the SE image, but with a small bounding box centered _**near**_ that x, y and with requested size for the psf image (what is the exact right bbox?)
- for the coadd, use the same wcs as the image coadd but again create a bbox of the appropriate size
- use a center near the coadd center (what bbox exactly?)
- use the warping code
```python
wexp = warper.warpExposure(
    psf_coadd_wcs,  # same as coadd wcs
    psf_exp,
    maxBBox=psf_exp.getBBox(),
    destBBox=psf_coadd_bbox,  # on center of coadd
)
```

**Comment by esheldon** [comment]
Care must be taken so that the coadded psf is centered

**Comment by beckermr** [comment]
Here is how I compute the PSF stamp params for DES: https://github.com/beckermr/pizza-cutter/blob/66b9e443f840798996b659a4f6ce59930681c776/pizza_cutter/des_pizza_cutter/_se_image.py#L708

For the coadd, the bounding box is just the patch plus a buffer if it does not already have one.

**Comment by beckermr** [comment]
For the coadd PSF wcs, find the pixel center nearest the center of the coadd and use that as the location to draw PSFs and as the location to coadd them too. For an odd psf stamp size, (xpix - (dim-1)/2, ypix-(dim-1)/2) is the correctlocation for the lower left corner of the bbox.

**Comment by beckermr** [comment]
```
        # we center the PSF at the nearest pixel center near the patch center
        col, row = wcs.sky2image(object_data['ra'][i], object_data['dec'][i])
        # this col, row includes the position offset
        # we don't need to remove it when putting them back into the WCS
        # but we will remove it later since we work in zero-indexed coords
        col = int(np.floor(col + 0.5))
        row = int(np.floor(row + 0.5))
        # ra, dec of the pixel center
        ra_psf, dec_psf = wcs.image2sky(col, row)
```

^ this is how I compute the location to draw the PSF. It is not the center of the slice coordinate system, it is the pixel center closest to the center of the slice coordinate system

**Comment by esheldon** [comment]
I tried something but I think I don't understand how the bboxes work, I get an empty psf

**Comment by beckermr** [comment]
This should be possible. We just don't know what to call.

File: descwl_coadd/coadd.py [hunk 7]
```diff
@@ -562,54 +568,82 @@ def get_psf_exp(
...
+        geom.Point2I(xmin + dim-1, ymin + dim-1),
+    )
+
+
+def get_coadd_psf_bbox(x, y, dim):
+    """
+    suggested by Matt Becker
+    """
+    xpix = int(x)
+    ypix = int(y)
```
Comment (beckermr) [inline descwl_coadd/coadd.py:638]: This only works for integer positions.

Comment (esheldon) [reply descwl_coadd/coadd.py:638]: how would you do it?

(note this isn't the cause of the image being all NaN)

Comment (beckermr) [reply descwl_coadd/coadd.py:638]: The coadd PSF position should always be integers, so the code is fine.

**Comment by beckermr** [comment]
When you read a patch from the butler you get back an image with this kind of WCS.

I wonder if the coadding code is relying not on the stated image bounds but on the shape of the array.

**Comment by esheldon** [comment]
My guess is I'm doing something that doesn't make any sense, but I'm not sure where to start since the codes are not documented

File: descwl_coadd/tests/test_coadd.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        epochs_per_band=epochs_per_band,
+        g1=0.02,
+        g2=0.00,
+        bands=bands,
+        psf=psf,
+        dither=dither,
+        rotate=dither,
+    )
+
+'''
```
Comment (beckermr) [inline descwl_coadd/tests/test_coadd.py:64]: This needs to be turned back on or deleted.

File: descwl_coadd/coadd.py [hunk 9]
```diff
@@ -1331,202 +793,85 @@ def _finish_init(self):
...
         noise[w] = 0.0
 
 
-def flag_bright_as_sat(*, mask):
+def flag_bright_as_sat(exp):
     """
     flag BRIGHT also as SAT so no detections will occur there
 
     we currently pull the bitmask value from the descwl_shear_sims
     package
```
Comment (beckermr) [inline descwl_coadd/coadd.py:861]: This comment is wrong now right?

File: descwl_coadd/coadd.py [hunk 10]
```diff
@@ -1331,202 +793,85 @@ def _finish_init(self):
...
-    print('pixels for badcols:', w[0].size)
-    if w[0].size > 0:
-        noise_exp.image.array[w] = exp.image.array[w]
+    logger = lsst.log.getLogger(name)
+    logger.setLevel(getattr(lsst.log, loglevel.upper()))
+    return logger
 
 
-def zero_bits(*, image, noise, mask, flags):
+def zero_bits(image, noise, mask, flags):
```
Comment (beckermr) [inline descwl_coadd/coadd.py:848]: Doc string here or add an _ to the function name?

File: descwl_coadd/coadd.py [hunk 11]
```diff
@@ -1331,202 +793,85 @@ def _finish_init(self):
...
+    coadd_pixcen, coadd_skycen = get_coadd_center(
+        coadd_wcs=coadd_wcs,
+        coadd_bbox=coadd_bbox,
+    )
+    coadd_cd_matrix = coadd_wcs.getCdMatrix(coadd_pixcen)
 
-    TODO: get a realistic value for the real data
-    """
+    # coadd_skycen = coadd_wcs.getSkyOrigin()
+    # coadd_cd_matrix = coadd_wcs.getCdMatrix(coadd_wcs.getPixelOrigin())
```
Comment (beckermr) [inline descwl_coadd/coadd.py:830]: Remove commented code

File: descwl_coadd/coadd.py [hunk 12]
```diff
@@ -1331,202 +793,85 @@ def _finish_init(self):
...
+        flags_for_maskfrac = self.coadd_exp.mask.getPlaneBitMask('BRIGHT')
+        self.meta['mask_frac'] = get_masked_frac(
+            mask=self.ormask,
+            flags=flags_for_maskfrac,
         )
-    )
 
 
-def make_stack_wcs(wcs):
+def get_coadd_psf_wcs(coadd_wcs, coadd_bbox, psf_dims):
```
Comment (beckermr) [inline descwl_coadd/coadd.py:803]: This routine appears wrong for the same reason as before. The coadd PSF WCS is the coadd WCS but with the PSF stamp labeled with the right bounds.

File: descwl_coadd/coadd.py [hunk 13]
```diff
@@ -1247,8 +708,9 @@ def _get_jac(self, *, cenx, ceny):
             Center for the output ngmix jacobian (not place of evaluation)
         """
 
-        coadd_cen = self.coadd_wcs.getPixelOrigin()
-        dm_jac = self.coadd_wcs.linearizePixelToSky(coadd_cen, geom.arcseconds)
+        coadd_wcs = self.coadd_exp.getWcs()
+        coadd_cen = coadd_wcs.getPixelOrigin()
+        dm_jac = coadd_wcs.linearizePixelToSky(coadd_cen, geom.arcseconds)
```
Comment (beckermr) [inline descwl_coadd/coadd.py:713]: If I understand it correctly, the coadd is a small subset of a larger WCS. This routine appears to return the jacobian at the center of the larger WCS, but we want it at the center of the smaller coadd within the larger system.

File: descwl_coadd/coadd.py [hunk 14]
```diff
@@ -1331,202 +793,85 @@ def _finish_init(self):
             store_pixels=False,
         )
```
Comment (beckermr) [inline descwl_coadd/coadd.py:794]: Why is the bmask zero? Why aren't we setting the ormask to zero and using the mask in the coadd as the bmask?

File: descwl_coadd/coadd.py [hunk 15]
```diff
@@ -1266,7 +728,8 @@ def _get_coadd_psf_obs(self):
         get the psf observation
         """
 
-        coadd_cen = self.coadd_wcs.getPixelOrigin()
+        coadd_wcs = self.coadd_exp.getWcs()
+        coadd_cen = coadd_wcs.getPixelOrigin()
```
Comment (beckermr) [inline descwl_coadd/coadd.py:732]: This is the wrong center. We want the center of the coadd slice image.

File: descwl_coadd/coadd.py [hunk 16]
```diff
@@ -1,1236 +1,697 @@
...
+    flags2interp = exp.mask.getPlaneBitMask(FLAGS2INTERP)
+    iimage, inoise = interpolate_image_and_noise(
+        image=exp.image.array,
+        noise=noise_exp.image.array,
+        weight=weight,
+        bmask=exp.mask.array,
+        bad_flags=flags2interp,
+    )
+    if iimage is None:
+        return None, None
```
Comment (beckermr) [inline descwl_coadd/coadd.py:313]: This should return three None's to match the return below.

File: descwl_coadd/coadd.py [hunk 17]
```diff
@@ -1,1236 +1,697 @@
...
+    exp_or_ref: afw_image.ExposureF or DeferredDatasetHandle
+        The input exposure, possible deferred
 
-    def _make_coadd(self, *, wexps, weights, input_recorder, psf_config=None):
-        """
-        make a coadd from warp images, as well as psf coadd
-        """
+    Returns
+    -------
+    exp, noise_exp
```
Comment (beckermr) [inline descwl_coadd/coadd.py:280]: Doc string is wrong.

File: descwl_coadd/coadd.py [hunk 18]
```diff
@@ -1,1236 +1,697 @@
...
-        warp_config.warpingKernelName = self.interp
-        warper = afw_math.Warper.fromConfig(warp_config)
+    Returns
+    -------
+    ExpsureF
+    """
+    coadd_exp = afw_image.ExposureF(coadd_bbox, coadd_wcs)
+    coadd_exp.mask.addMaskPlane("REJECTED")
+    coadd_exp.mask.addMaskPlane("CLIPPED")
+    coadd_exp.mask.addMaskPlane("SENSOR_EDGE")
```
Comment (beckermr) [inline descwl_coadd/coadd.py:230]: I didn't see theser flags being used anywhere? Did I miss it?

**Review by beckermr — CHANGES_REQUESTED** [review]
I left more comments on WCS handling. We should only use Jacobian approximations right at the end. Further, it appears that for the coadd PSF you are making a new WCS instead of using the coadd WCS with the PSF box defined to overlap the center of the slice.

File: descwl_coadd/tests/test_coadd.py (hunk 8 above)
Comment (esheldon) [reply descwl_coadd/tests/test_coadd.py:64]: whoops

File: descwl_coadd/coadd.py (hunk 9 above)
Comment (esheldon) [reply descwl_coadd/coadd.py:861]: will fix

File: descwl_coadd/coadd.py (hunk 10 above)
Comment (esheldon) [reply descwl_coadd/coadd.py:848]: will add doc

File: descwl_coadd/coadd.py (hunk 11 above)
Comment (esheldon) [reply descwl_coadd/coadd.py:830]: whoops

File: descwl_coadd/coadd.py (hunk 12 above)
Comment (esheldon) [reply descwl_coadd/coadd.py:803]: no longer used, will remove

File: descwl_coadd/coadd.py (hunk 15 above)
Comment (esheldon) [reply descwl_coadd/coadd.py:732]: true, and I will fix

but we are storing a fixed kernel so it doesn't matter

it matters more for the jacobian a few lines up

File: descwl_coadd/coadd.py (hunk 16 above)
Comment (esheldon) [reply descwl_coadd/coadd.py:313]: will fix

File: descwl_coadd/coadd.py (hunk 17 above)
Comment (esheldon) [reply descwl_coadd/coadd.py:280]: will fix

File: descwl_coadd/coadd.py (hunk 18 above)
Comment (esheldon) [reply descwl_coadd/coadd.py:230]: DM adds these planes in their coadd processing, so I added them to make the default set of bits the same

File: descwl_coadd/coadd.py (hunk 14 above)
Comment (esheldon) [reply descwl_coadd/coadd.py:794]: I don't think we are actually using these downstream, we use the mask attribute.

And we use that mask attribute as the ormask, so there is some consistency here.

We need to add the mask frac that you use for DES

Comment (esheldon) [reply descwl_coadd/coadd.py:794]: I mean "mask attribute of the exposure attribute"

Comment (esheldon) [reply descwl_coadd/coadd.py:794]: see issue #48

File: descwl_coadd/coadd.py (hunk 18 above)
Comment (beckermr) [reply descwl_coadd/coadd.py:229]: Maybe add a comment to that effect?

Comment (esheldon) [reply descwl_coadd/coadd.py:231]: done

File: descwl_coadd/coadd.py (hunk 14 above)
Comment (esheldon) [reply descwl_coadd/coadd.py:805]: Can we punt on this and address it in #48?

Comment (beckermr) [reply descwl_coadd/coadd.py:805]: ofc!

**Review by beckermr — APPROVED** [review]
I think we are good to go. We still need someone from DM to look at the code carefully. Maybe Eli?

**Pull request merged**